File: src/main/java/dsl/FunctionBuilder.java
package dsl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FunctionBuilder {
    private final String name;
    private final List<String> parameters = new ArrayList<>();

    FunctionBuilder(String name) {
        this.name = name;
    }

    public FunctionParametersBuilder with(String... params) {
        parameters.addAll(Arrays.asList(params));
        return new FunctionParametersBuilder(name, parameters);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/functions/FunctionBuilder.kt
package xyz.mahmoudahmed.dsl.functions

/**
 * Builder for creating a new mapping function.
 */
class FunctionBuilder(private val name: String) {
    private val parameters = mutableListOf<String>()

    /**
     * Defines the parameters for the function.
     *
     * @param params The parameter names
     * @return Builder for configuring the function parameters
     */
    fun with(vararg params: String): FunctionParametersBuilder {
        parameters.addAll(params)
        return FunctionParametersBuilder(name, parameters)
    }

    /**
     * Creates a function with no parameters.
     *
     * @return Builder for configuring the function body
     */
    fun withNoParameters(): FunctionBodyBuilder {
        return FunctionBodyBuilder(name, emptyList()) { it[0] }
    }
}

